internal/impl/pulsar/output.go

```diff
@@ -198,6 +232,57 @@ func (p *pulsarWriter) Write(ctx context.Context, msg *service.Message) error {
 	return err
 }
 
+func (p *pulsarWriter) WriteBatch(ctx context.Context, batch service.MessageBatch) error {
+	var r pulsar.Producer
+	p.m.RLock()
+	if p.producer != nil {
+		r = p.producer
+	}
+	p.m.RUnlock()
+
+	if r == nil {
+		return component.ErrNotConnected
+	}
+
+	msgs := make([]*pulsar.ProducerMessage, len(batch))
+
+	err := batch.WalkWithBatchedErrors(func(i int, msg *service.Message) error {
+		b, err := msg.AsBytes()
+		if err != nil {
+			return err
+		}
+
+		m := &pulsar.ProducerMessage{
+			Payload: b,
+		}
+		if key := p.key.Bytes(msg); len(key) > 0 {
+			m.Key = string(key)
+		}
+		if orderingKey := p.orderingKey.Bytes(msg); len(orderingKey) > 0 {
+			m.OrderingKey = string(orderingKey)
+		}
+
+		msgs[i] = m
```

Comment:
Mainly because I'd like to put `SendAsync` and `Flush` as close as possible. In case we fail anywhere during `WalkWithBatchedErrors`, we don't need to call `SendAsync`

Reply:
The only error could happen here is `msg.AsBytes()` which should rarely happen ( messages should always be able to be presented as bytes ). And even it fails, I think it's fine too, because any `SendAsync` could fail as well, it's not guaranteed that all messages in a batch will be sent anyways. So I think it's fine to call SendAsync here.